torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py

```diff
@@ -318,130 +325,139 @@ def _annotate_conv_relu(
+    return a list of annotated partitions.
 
-        weight = conv_node.args[1]
-        assert isinstance(weight, Node)
-        input_qspec_map[weight] = get_weight_qspec(quantization_config)
+    The output of the pattern must include a dictionary from string name to node
+    for the following names: "input", "conv", "weight", "bias", and "output".
+    """
+    def get_pattern(conv_fn: Callable):
+        def _conv_bn(x, conv_weight, conv_bias, bn_weight, bn_bias, bn_rm, bn_rv):
+            conv = conv_fn(x, conv_weight, conv_bias)
+            bn = F.batch_norm(conv, bn_rm, bn_rv, bn_weight, bn_bias, training=True)
+            if has_relu:
+                output = F.relu(bn)
+            else:
+                output = bn
+            return output, {
+                "input": x,
+                "weight": conv_weight,
+                "bias": conv_bias,
+                "output": output,
+            }
+        return _conv_bn
+
+    # Needed for matching, otherwise the matches gets filtered out due to unused
+    # nodes returned by batch norm
+    gm.graph.eliminate_dead_code()
+    gm.recompile()
+
+    matches = []
+    combinations = [
+        (F.conv2d, _conv2d_bn_example_inputs),
+    ]
 
-        bias = conv_node.args[2] if len(conv_node.args) > 2 else None
-        if isinstance(bias, Node):
-            input_qspec_map[bias] = get_bias_qspec(quantization_config)
+    # Add cuda dimension
+    new_combinations = [(x, y, False) for (x, y) in combinations]
+    if torch.cuda.is_available():
+        new_combinations.extend([(x, y, True) for (x, y) in combinations])
+    combinations = new_combinations
+
+    # Match against all conv dimensions and cuda variants
+    for conv_fn, example_inputs, is_cuda in combinations:
+        pattern = get_aten_graph_module(get_pattern(conv_fn), example_inputs, is_cuda)
+        pattern.graph.eliminate_dead_code()
+        pattern.recompile()
+        matcher = SubgraphMatcherWithNameNodeMap(pattern, ignore_literals=True)
+        matches.extend(matcher.match(gm.graph))
+
+    # Annotate nodes returned in the matches
+    annotated_partitions = []
+    for match in matches:
+        name_node_map = match.name_node_map
+        input_node = name_node_map["input"]
+        weight_node = name_node_map["weight"]
+        bias_node = name_node_map["bias"]
+        output_node = name_node_map["output"]
+
+        # Find conv node and validate its args
+        if len(input_node.users) != 1:
+            raise ValueError("Expected input node to be consumed by conv only")
+        if len(weight_node.users) != 1:
+            raise ValueError("Expected weight node to be consumed by conv only")
+        input_user = list(input_node.users.keys())[0]
+        weight_user = list(weight_node.users.keys())[0]
+        if weight_user is not input_user:
+            raise ValueError("Expected weight user to be the same as input user")
+        conv_node = input_user
```

Comment:
This still assumes conv_node is not decomposed. (when it's decomposed, input_user, weight_user and bias_user could be different nodes, e.g. linear is a good example)

I think you'll need to annotate `input_user`, `weight_user` and `bias_user` separately to be 100% robust to decompositions of the conv op

Reply:
As discussed offline, we will just use the `conv_node` returned from the subgraph matcher here and just validate the conv args for now. This is because we can't always rely on the first user of the input node, for example in resnet18 where there are skip connections